gem/optimise.py

```diff
@@ -413,3 +474,403 @@ def aggressive_unroll(expression):
+        for to_delete in factored_out:
+            self.rep.remove(to_delete)
+        new_summand = OrderedDict()
+        if self._decide_key(node) == key:
+            new_summand[key] = [mcf, node]
+        else:
+            new_summand[key] = [mcf]
+            new_summand[self._decide_key(node)] = [node]
+        self.rep.append(new_summand)
+        self.factorise_key(key)  # Continue factorising
+
+    def factorise_arg(self, factors_seq):
+        """
+        Factorise sequentially with common factors from :param factors_seq
+        :param factors_seq: sequence of factors used to factorise
+        """
+        if not factors_seq:
+            # TODO: investigate if worthwhile to do a pass to check const common factors here
+            # self.factorise_key('other')
+            # self.factorise_key('const')
+            return self.generate_node()
+        cf = factors_seq[0]  # pick the first common factor
+        key = self._decide_key(cf)
+        factored_out = list()
+        # cf * Sum(_summands), where summand = Product(_factors)
+        _summands = list()
+        for summand in self.rep:
+            if key not in summand:
+                continue
+            factors = summand[key]
+            if len(factors) > 1:
+                raise AssertionError("There should be only one argument factor")
+            if cf not in factors:
+                continue
+            factored_out.append(summand)  # mark for deleting later
+            # TODO: default fields might be missing from this OrderedDict(), consider wrap an object around it
+            _summands.append(OrderedDict([(_k, _v) for (_k, _v) in iteritems(summand) if _k != key]))
+        self.opt_node = None
+        if len(_summands) > 1:
+            # Proceed with the next common factor for the factorised part
+            lo = LoopOptimiser(node=None, arg_ind=self.arg_ind, rep=_summands)
+            lo.factorise_arg(factors_seq[1:])
+            node = lo.generate_node()
+            # Delete factored out lines in rep
+            for to_delete in factored_out:
+                self.rep.remove(to_delete)
+            # Create new line in rep
+            # TODO: Need an object to wrap around summands
+            new_summand = OrderedDict()
+            for i in ['const', 'other'] + list(self.arg_ind_flat):
+                new_summand[i] = list()
+            new_summand[self._decide_key(cf)] = [cf]
+            new_summand[self._decide_key(node)] = [node]
+            self.rep.append(new_summand)
+        # Proceed with the next common factor
+        lo = LoopOptimiser(node=None, arg_ind=self.arg_ind, rep=self.rep)
+        lo.factorise_arg(factors_seq[1:])
+        return lo.generate_node()
+
+    def generate_node(self):
+        # TODO: Here need to consider the order of forming products, currently this only ensures same group gets multiplied first. Consider lexico order between groups.
+        if self.opt_node:
+            return self.opt_node
```

Comment:
This weird caching is an anti pattern as well.  A `LoopOptimiser` can always generate a node, but if I call it at the wrong moment, then do some optimisation and then call `generate_node` again, I will get the wrong thing (because the modifications of `self.rep` won't have been noticed).